Add bang_olufsen integration

mj23000 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Add the bang_olufsen integration for Bang & Olufsen Mozart devices. This integration has been in the works for a long time and has been tested publicly at https://github.com/bang-olufsen/bangolufsen-hacs 

PR to add brand icons to core_integrations in brands repo: https://github.com/home-assistant/brands/pull/4605

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/30432

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

mj23000 committed:
Add bangolufsen integration

mj23000 committed:
add untested files to .coveragerc

mj23000 committed:
Merge branch 'home-assistant:dev' into bangolufsen

mj23000 committed:
Simplify integration to media_player platform

mj23000 committed:
Merge branch 'dev' into bangolufsen

mj23000 committed:
Remove missing files from .coveragerc

mj23000 committed:
Merge branch 'dev' into bangolufsen

mj23000 committed:
Merge branch 'dev' into bangolufsen

mj23000 committed:
Merge branch 'dev' into bangolufsen

mj23000 committed:
Merge branch 'dev' into bangolufsen

mj23000 committed:
Add beolink_set_relative_volume custom service
Tweaks

mj23000 committed:
Merge branch 'dev' into bangolufsen

mj23000 committed:
Remove custom services
Remove grouping as it was dependent on custom services

mj23000 committed:
Merge branch 'dev' into bangolufsen

mj23000 committed:
Merge branch 'dev' into bangolufsen

joostlek commented on tests/components/bangolufsen/const.py:
```diff
@@ -0,0 +1,189 @@
+    TEST_VOLUME_STEP = 5
+
+
+class TestConstantsConfigFlow(TestConstants):
```
Please remove the classes from this file. You can create different test configurations using test fixtures.

mj23000 replied:
I've revamped the testing to (hopefully) properly use fixtures

joostlek commented on tests/components/bangolufsen/util.py:
```diff
@@ -0,0 +1,17 @@
+from tests.common import MockConfigEntry
+
+
+def mock_entry() -> MockConfigEntry:
```
Ah you do have some kind of fixture. Can you transform this into a fixture? Also move this to `conftest.py`

mj23000 replied:
Done

joostlek reviewed (dismissed):
Initial review. I am no expert on integrations using websockets so I skimmed over that part but this PR is way too big in its current form. Please cut down in size to speed up the process!

mj23000 committed:
Merge branch 'dev' into bangolufsen

mj23000 committed:
Update API to 3.2.1.150.0
Reduce and optimize code with feedback from joostlek
Tweaks

mj23000 committed:
Updated testing

mj23000 commented:
> Initial review. I am no expert on integrations using websockets so I skimmed over that part but this PR is way too big in its current form. Please cut down in size to speed up the process!

Thank you very much for the feedback. I've tried to address most of the points that you made and also reduced the code further. Sorry for the slow communication.

Kind regards

mj23000 commented on homeassistant/components/bangolufsen/__init__.py:
```diff
@@ -0,0 +1,80 @@
+
+    # If connection can't be made abort.
+    if not await init_entities(hass, entry):
+        return False
```
Yes I've done this now.

mj23000 commented on homeassistant/components/bangolufsen/config_flow.py:
```diff
@@ -0,0 +1,278 @@
+        return BangOlufsenOptionsFlowHandler(config_entry)
+
+
+class BangOlufsenOptionsFlowHandler(OptionsFlow):
```
I've removed it for now.

mj23000 commented on homeassistant/components/bangolufsen/const.py:
```diff
@@ -0,0 +1,372 @@
+        )
+
+
+class BangOlufsenEntity(Entity, BangOlufsenVariables):
```
Done

mj23000 commented on homeassistant/components/bangolufsen/device_trigger.py:
```diff
@@ -0,0 +1,199 @@
+"""Device triggers for the Bang & Olufsen integration."""
```
Done

mj23000 commented on homeassistant/components/bangolufsen/config_flow.py:
```diff
@@ -0,0 +1,278 @@
+
+        return await self.async_step_confirm()
+
+    async def async_step_confirm(
```
Yes, I've removed this part of the setup for now.

mj23000 commented on homeassistant/components/bangolufsen/config_flow.py:
```diff
@@ -0,0 +1,278 @@
+                min=MAX_VOLUME_RANGE.start,
+                max=(MAX_VOLUME_RANGE.stop - 1),
+            ),
+        ),
```
The NumberSelector is used now.

mj23000 commented on homeassistant/components/bangolufsen/config_flow.py:
```diff
@@ -0,0 +1,278 @@
+
+    VERSION = 1
+
+    async def _validate_host(self) -> None:
```
Yes, I've done this now.

mj23000 commented on homeassistant/components/bangolufsen/config_flow.py:
```diff
@@ -0,0 +1,278 @@
+            self._serial_number = beolink_self.jid.split(".")[2].split("@")[0]
+
+        except ApiException as error:
+            raise AbortFlow(reason=API_EXCEPTION) from error
```
Done.

mj23000 commented on homeassistant/components/bangolufsen/config_flow.py:
```diff
@@ -0,0 +1,278 @@
+            data[CONF_HOST] = self._host
+            data[CONF_MODEL] = self._model
+            data[CONF_BEOLINK_JID] = self._beolink_jid
+            data[CONF_NAME] = self._name
```
I've updated the code to use the title instead of CONF_NAME.

mj23000 commented on homeassistant/components/bangolufsen/const.py:
```diff
@@ -0,0 +1,372 @@
+class ArtSizeEnum(Enum):
+    """Enum used for sorting images that have size defined by a string."""
+
+    small = 1
```
I've renamed the Enum classes to follow this casing.

joostlek commented:
Feel free to resolve comments if you think you solved it btw :)

mj23000 commented on homeassistant/components/bangolufsen/const.py:
```diff
@@ -0,0 +1,372 @@
+    """Enum for accessing and storing the entities in hass."""
+
+    WEBSOCKET = "websocket"
+    MEDIA_PLAYER = "media_player"
```
In the "full" version of the integration, this Enum has 9 different members. I think it makes sense when the Enum gets that many members.

mj23000 commented on homeassistant/components/bangolufsen/const.py:
```diff
@@ -0,0 +1,372 @@
+    BEOLINK_JOIN_RESULT: Final[str] = "beolink_join_result"
+    BEO_REMOTE_BUTTON: Final[str] = "beo_remote_button"
+    BUTTON: Final[str] = "button"
+    CURTAINS: Final[str] = "curtains"
```
Some Bang & Olufsen devices have moveable covers [Beolab 28](https://www.bang-olufsen.com/en/dk/speakers/beolab-28) that called "curtains". This is currently not used so the constant has been removed

mj23000 commented on homeassistant/components/bangolufsen/const.py:
```diff
@@ -0,0 +1,372 @@
+DOMAIN: Final[str] = "bangolufsen"
+
+# Default values for configuration.
+DEFAULT_HOST: Final[str] = "192.168.1.1"
```
This was mostly to save time in testing. It has been removed now.

mj23000 commented on homeassistant/components/bangolufsen/media_player.py:
```diff
@@ -0,0 +1,925 @@
+    def __init__(self, entry: ConfigEntry) -> None:
+        """Initialize the media player."""
+        MediaPlayerEntity.__init__(self)
+        BangOlufsenEntity.__init__(self, entry)
```
Done

mj23000 commented on homeassistant/components/bangolufsen/media_player.py:
```diff
@@ -0,0 +1,925 @@
+        BangOlufsenEntity.__init__(self, entry)
+
+        # Static entity attributes
+        self._attr_device_class = MediaPlayerDeviceClass.SPEAKER
```
It looks like it can't be moved outside the constructor. The device class was not set if it was.

mj23000 commented on homeassistant/components/bangolufsen/media_player.py:
```diff
@@ -0,0 +1,925 @@
+        self._attr_icon = "mdi:speaker-wireless"
+        self._attr_supported_features = BANGOLUFSEN_FEATURES
+        self._attr_unique_id = self._unique_id
+        self._attr_should_poll = True
```
The BangOlufsenEntity has it disabled by default. The WebSocket notification channel is only used for some immediate state information. The endpoint that is polled is not provided through the WebSocket notification channel.

mj23000 commented on homeassistant/components/bangolufsen/media_player.py:
```diff
@@ -0,0 +1,925 @@
+    @property
+    def state(self) -> MediaPlayerState:
+        """Return the current state of the media player."""
+        return cast(MediaPlayerState, StateEnum[self._state])
```
I've replaced the Enum with a dict

mj23000 commented on homeassistant/components/bangolufsen/media_player.py:
```diff
@@ -0,0 +1,925 @@
+            content_filter=lambda item: item.media_content_type.startswith("audio/"),
+        )
+
+    # Custom services:
```
Done

mj23000 commented on homeassistant/components/bangolufsen/services.yaml:
```diff
@@ -0,0 +1,190 @@
+beolink_join:
```
Looks like i forgot to remove the services.yaml file. It's removed now

mj23000 commented on homeassistant/components/bangolufsen/media_player.py:
```diff
@@ -0,0 +1,925 @@
+            async_req=True
+        ).get()
+
+        _LOGGER.info(
```
Logging is now done on debug.

mj23000 committed:
Remove unused options schema

mj23000 committed:
Fix bugfix setting wrong state

mj23000 committed:
Fix wrong initial state

mj23000 committed:
Bump API

mj23000 committed:
Merge branch 'dev' into bangolufsen

mj23000 committed:
Merge branch 'dev' into bangolufsen

mj23000 committed:
Merge branch 'dev' into bangolufsen

mj23000 committed:
Fix Beosound Level not reconnecting properly

mj23000 committed:
Remove unused constant

mj23000 committed:
Fix wrong variable checked to determine source

mj23000 committed:
Merge branch 'dev' into bangolufsen

emontnemery commented on homeassistant/components/bangolufsen/__init__.py:
```diff
@@ -0,0 +1,83 @@
+    if not entry.unique_id:
+        raise ConfigEntryError("Can't retrieve unique id from config entry. Aborting")
+
+    # If connection can't be made abort.
```
Why do we abort here? Since this is a media player I would expect it to be a very common case that the media player is powered off, it seems reasonable that the media player entity is still created in that case, but unavailable?

mj23000 replied:
Yes, this works as expected.

emontnemery commented on homeassistant/components/bangolufsen/__init__.py:
```diff
@@ -0,0 +1,83 @@
+    media_player = BangOlufsenMediaPlayer(entry)
+
+    # Add the created entities
+    hass.data[DOMAIN][entry.unique_id] = {
+        ENTITY_ENUM.WEBSOCKET: websocket,
+        ENTITY_ENUM.MEDIA_PLAYER: media_player,
+    }
```
Please use a dataclass instead of a dict.

mj23000 replied:
Done.

emontnemery commented on homeassistant/components/bangolufsen/__init__.py:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    # Start the WebSocket listener with a delay to allow for entity and dispatcher listener creation
+    async_call_later(hass, WEBSOCKET_CONNECTION_DELAY, websocket.connect_websocket)
```
A delay is never an acceptable way to wait for other things to happen. Please implement the needed synchronization instead.

mj23000 replied:
How would you do this? The issue is that not all of the entities are created fast enough and therefore have not  connected all dispatchers listeners before the websocket sends the dispatcher events.

The integration does not heavily depend on information from the websocket notification channel upon startup right now (at this point since only the media_player platform is included), so I've simply removed the delay for now.

emontnemery replied:
There are many ways to synchronize the startup, I'm not sure what best suits your needs.

If you want to ensure that the entities have been added to Home Assistant before the websocket is connected you could let each platform's (only `media_player` for now) `async_setup_entry` append a key to a "waiting list" stored in `BangOlufsenData` for each entity object they create. The entities should then pop themselves from this list when either of `async_added_to_hass` or `async_will_remove_from_hass` are called. `async_will_remove_from_hass` will be called if the user has disabled an entity, but also if the config entry is unloaded, so you need to handle `async_will_remove_from_hass` being called after `async_added_to_hass` has been called.

Once the list is empty, call `websocket.connect_websocket`.

emontnemery commented on homeassistant/components/bangolufsen/config_flow.py:
```diff
@@ -0,0 +1,203 @@
+    jid: str
+
+
+class BangOlufsenConfigFlowHandler(ConfigFlow, domain=DOMAIN):
+    """Handle a config flow."""
+
+    def __init__(self) -> None:
+        """Init the config flow."""
+        self._host: str = ""
+        self._name: str = ""
+        self._model: str = ""
+        self._serial_number: str = ""
+        self._beolink_jid: str = ""
+
+        self._client: MozartClient | None = None
+
```
Since the initialization values don't depend on arguments passed to __init__, move them to class level.
If the strings can only be strings, there's no need to explicitly state that.
To avoid useless checking of `_client`, we can unconditionally create the the client in the user and confirm steps and just add a type hint at class level

```suggestion
class BangOlufsenConfigFlowHandler(ConfigFlow, domain=DOMAIN):
    """Handle a config flow."""
    _beolink_jid = ""
    _client: MozartClient
    _host = ""
    _name = ""
    _model = ""
    _serial_number = ""

    def __init__(self) -> None:
        """Init the config flow."""

```

mj23000 replied:
I've done this.

emontnemery commented on homeassistant/components/bangolufsen/config_flow.py:
```diff
@@ -0,0 +1,203 @@
+
+    async def _compile_data(self) -> UserInput:
+        """Compile data for entry creation."""
+        if not self._client:
+            self._client = MozartClient(self._host, urllib3_logging_level=logging.ERROR)
```
```suggestion
```

emontnemery commented on homeassistant/components/bangolufsen/config_flow.py:
```diff
@@ -0,0 +1,203 @@
+            self._client = MozartClient(self._host, urllib3_logging_level=logging.ERROR)
+
+        # Get current volume settings
+        volume_settings = cast(
+            ApplyResult[VolumeSettings],
+            self._client.get_volume_settings(async_req=True),
+        ).get()
```
The library has a `py.typed` marker, why do we need to cast?

emontnemery replied:
Also, does `ApplyResult.get()` block?

mj23000 replied:
The "async_req=True" parameter changed the output type of the method, but the typing did not account for this.
And also yes it was blocking. I've done a big API overhaul to fix this and the typing.

emontnemery commented on homeassistant/components/bangolufsen/config_flow.py:
```diff
@@ -0,0 +1,203 @@
+        data[CONF_HOST] = self._host
+        data[CONF_MODEL] = self._model
+        data[CONF_BEOLINK_JID] = self._beolink_jid
+        data[CONF_VOLUME_STEP] = DEFAULT_VOLUME_STEP
```
We are likely to no longer allow media players to implement both support for setting volume to a certain level and custom volume_up / volume_down. Hence, I strongly advice to remove this, as well as the implementation of volume_up and volume_down in the media player entity

mj23000 replied:
I've removed volume step entirely.

emontnemery commented on homeassistant/components/bangolufsen/config_flow.py:
```diff
@@ -0,0 +1,203 @@
+            self._client = MozartClient(self._host, urllib3_logging_level=logging.ERROR)
+
+        # Get current volume settings
+        volume_settings = cast(
+            ApplyResult[VolumeSettings],
+            self._client.get_volume_settings(async_req=True),
+        ).get()
```
Are the volume settings something the user can configure via some Bang & Olufsen service? If yes, it seems weird to fetch this once when creating the config entry and then never update it.

mj23000 replied:
volume settings are configurable through the B&O app and previously through config_flow options (where it was also updated).

emontnemery replied:
Right, so what's the expected behavior if the user changes the volume settings from the B&O app after the config entry has been added? As I understand it, Home Assistant will now ignore those changes because the volume settings are only read once, during creation of the config entry?

mj23000 replied:
Yes that's right. These volume options seem to be mostly useless (Again, configurable through the app) and not used in the integration, so I should probably just remove them.

mj23000 replied:
Volume settings have been removed.

emontnemery commented on homeassistant/components/bangolufsen/config_flow.py:
```diff
@@ -0,0 +1,203 @@
+        }
+
+        await self.async_set_unique_id(self._serial_number)
+        self._abort_if_unique_id_configured()
```
Please make sure to update the IP-address of an existing config entry as done here: https://github.com/home-assistant/core/pull/90511

mj23000 replied:
I've done this.

emontnemery commented on homeassistant/components/bangolufsen/config_flow.py:
```diff
@@ -0,0 +1,203 @@
+        self, user_input: UserInput | None = None
+    ) -> FlowResult:
+        """Confirm the configuration of the device."""
+        if user_input is not None:
+            data = await self._compile_data()
+            return self.async_create_entry(title=self._name, data=data)
```
```suggestion
        if user_input is not None:
            self._client = MozartClient(self._host, urllib3_logging_level=logging.ERROR)
            data = await self._compile_data()
            return self.async_create_entry(title=self._name, data=data)
```

mj23000 replied:
I've created the client in the zeroconf method, to avoid the client being created twice in a manual config flow.

emontnemery replied:
Right. To avoid wasting memory if the user is not interested in adding their B&O speaker to Home Assistant, maybe add an intermediate step, `async_step_zeroconf_confirm` which creates the client and then calls `async_step_confirm`, or check if the client exists and create it if it doesn't in `_compile_data`.

emontnemery commented on homeassistant/components/bangolufsen/const.py:
```diff
@@ -0,0 +1,309 @@
+from homeassistant.helpers.device_registry import DeviceEntry
+
+
+class ART_SIZE_ENUM(Enum):
+    """Enum used for sorting images that have size defined by a string."""
+
+    small = 1
+    medium = 2
+    large = 3
```
Definitions like these don't seem relevant for the Home Assistant integration, and thus should be in a supporting API, not in Home Assistant. How come they're not in `mozart_api`?

mj23000 replied:
It seemed like clutter for the API client, but I guess it is also clutter for the Home Assistant integration. I've moved this functionality to `mozart_api`.

emontnemery commented on homeassistant/components/bangolufsen/const.py:
```diff
@@ -0,0 +1,372 @@
+NO_METADATA: Final[tuple] = (None, "", 0)
+
+
+def get_device(hass: HomeAssistant | None, unique_id: str) -> DeviceEntry | None:
```
I agree, please move this as well as `BangOlufsenVariables` out of here.
`get_device` could be moved to a `util` or `helper` module, `BangOlufsenVariables` to a `models` module or to the `entity` module.

emontnemery commented on homeassistant/components/bangolufsen/media_player.py:
```diff
@@ -0,0 +1,777 @@
+)
+
+
+PARALLEL_UPDATES = 0
```
Is this a limitation of `mozart_api`?

mj23000 replied:
I've removed this.

emontnemery commented on homeassistant/components/bangolufsen/media_player.py:
```diff
@@ -0,0 +1,777 @@
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up a Media Player entity from config entry."""
+    entity = hass.data[DOMAIN][config_entry.unique_id][ENTITY_ENUM.MEDIA_PLAYER]
```
As previously discussed, please create the media player entity here.

emontnemery commented on homeassistant/components/bangolufsen/media_player.py:
```diff
@@ -0,0 +1,777 @@
+        self._model: str = self.entry.data[CONF_MODEL]
+        self._volume_step: int = self.entry.data[CONF_VOLUME_STEP]
+
+        self._attr_device_class = MediaPlayerDeviceClass.SPEAKER
```
Move this to class level

mj23000 replied:
I can't move the self._model because it depends on the entry data and self._attr_device_class would cause the `BangOlufsenEntity` device class to be applied instead.

emontnemery commented on homeassistant/components/bangolufsen/media_player.py:
```diff
@@ -0,0 +1,777 @@
+            name=cast(str, self.name),
+        )
+        self._attr_name = self._name
+        self._attr_should_poll = True
```
Why is polling needed? Are we not guaranteed to get updates from the websocket API?
Also, move this to class level.

mj23000 replied:
Polling was needed becuase repeat / shuffle settings are not received through the WebSocket notification channel. I did however notice that async_update() method was being called too much, so I've this functionality for now.

emontnemery replied:
You can control how often `async_update()` is called by setting a `SCAN_INTERVAL` on module level

emontnemery commented on homeassistant/components/bangolufsen/media_player.py:
```diff
@@ -0,0 +1,777 @@
+        self._last_update: datetime = datetime(1970, 1, 1, 0, 0, 0, 0)
+        self._media_image: Art = Art()
+        self._queue_settings: PlayQueueSettings = PlayQueueSettings()
+        self._remote_leader: BeolinkLeader | None = None
```
This is never set. Please don't add support for potential future features.

mj23000 replied:
I've removed the last_update, queue_settings and remote leader.

emontnemery commented on homeassistant/components/bangolufsen/media_player.py:
```diff
@@ -0,0 +1,777 @@
+        # Misc. variables.
+        self._audio_sources: dict[str, str] = {}
+        self._beolink_listeners: list[BeolinkListener] = []
+        self._last_update: datetime = datetime(1970, 1, 1, 0, 0, 0, 0)
```
```suggestion
```

emontnemery commented on homeassistant/components/bangolufsen/config_flow.py:
```diff
@@ -0,0 +1,203 @@
+                beolink_self = cast(
+                    ApplyResult[BeolinkPeer],
+                    self._client.get_beolink_self(async_req=True, _request_timeout=3),
+                ).get()
```
Same question, does this block?

mj23000 replied:
Yes it did.

emontnemery commented on homeassistant/components/bangolufsen/media_player.py:
```diff
@@ -0,0 +1,777 @@
+                if self._playback_state.value:
+                    self._state = self._playback_state.value
+
+        self._last_update = utcnow()
```
```suggestion
        self._attr_media_position_updated_at = utcnow()
```

emontnemery commented on homeassistant/components/bangolufsen/media_player.py:
```diff
@@ -0,0 +1,777 @@
+    async def _update_playback_progress(self, data: PlaybackProgress) -> None:
+        """Update _playback_progress and last update."""
+        self._playback_progress = data
+        self._last_update = utcnow()
```
```suggestion
        self._attr_media_position_updated_at = utcnow()
```

emontnemery commented on homeassistant/components/bangolufsen/media_player.py:
```diff
@@ -0,0 +1,777 @@
+            return self._playback_progress.progress
+        return None
+
+    @property
+    def media_position_updated_at(self) -> datetime:
+        """Return the last time that the playback position was updated."""
+        return self._last_update
+
```
```suggestion
```

emontnemery commented on homeassistant/components/bangolufsen/entity.py:
```diff
@@ -0,0 +1,48 @@
+    def __init__(self, entry: ConfigEntry) -> None:
+        """Initialize the object."""
+        BangOlufsenVariables.__init__(self, entry)
+        self._dispatchers: list[Callable] = []
```
Remove this, and instead add the dispatchers like this:
```py
self.async_on_remove(
    async_dispatcher_connect(
        self.hass,
        f"{self._unique_id}_{CONNECTION_STATUS}",
        self._update_connection_state,
    )
)
```

emontnemery commented on homeassistant/components/bangolufsen/manifest.json:
```diff
@@ -0,0 +1,11 @@
+{
+  "domain": "bangolufsen",
```
Wouldn't bang_olufsen be a better fit?
```suggestion
  "domain": "bang_olufsen",
```

mj23000 replied:
This is simply a styling choice that we prefer, but it is not a dealbreaker if the latter is more "Home Assistant" like.

Sorry for the late response, I wanted to take time properly fix the code. I want to say thank you for the thorough code review. It really means a lot!

Kind regards

emontnemery replied:
`bang_olufsen` would indeed be more "Home Assistant" like, please change it 👍

mj23000 replied:
I've changed the name to bang_olufsen.

mj23000 committed:
Merge branch 'dev' into bangolufsen

mj23000 committed:
Update integration with feedback from emontnemery

mj23000 committed:
Update integration with feedback from emontnemery

mj23000 committed:
Merge branch 'dev' into bangolufsen

mj23000 committed:
Remove unused code

mj23000 commented on homeassistant/components/bangolufsen/__init__.py:
```diff
@@ -0,0 +1,83 @@
+
+    # Ensure that a unique id is available
+    if not entry.unique_id:
+        raise ConfigEntryError("Can't retrieve unique id from config entry. Aborting")
```
I agree. It has been removed.

mj23000 committed:
Move API client into dataclass
Fix not all config_flow exceptions caught
Tweaks

mj23000 committed:
Merge branch 'dev' into bangolufsen

mj23000 committed:
Add Bang & Olufsen brand

emontnemery commented on homeassistant/components/bangolufsen/__init__.py:
```diff
@@ -0,0 +1,67 @@
+    )
+
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+    websocket.connect_websocket()
```
The name of this method suggests it can fail if the device happened to go offline between the battery state check and now. If that's the case, we should catch the relevant errors and rethrow as `ConfigEntryNotReady`.

mj23000 replied:
Currently the connect_websocket() method simply hangs if a connection can't be made. I'll update the API to fix this and throw `ConfigEntryNotReady`.

mj23000 committed:
Revert "Add Bang & Olufsen brand"

This reverts commit 57b2722078ae0b563880306c6457d2cf3f528070.

mj23000 commented on homeassistant/brands/bangolufsen.json:
Got it. I've reverted the commit.

mj23000 committed:
Merge branch 'dev' into bangolufsen

emontnemery reviewed (changes requested):
This looks pretty good now 👍  Please address the new comments.

mj23000 committed:
Remove volume options from setup
Simplify device checks
rename integration to bang_olufsen
update tests to pass
Update API

mj23000 committed:
Remove _device from base
Add _device to websocket

mj23000 committed:
Move SW version device update to websocket
Sort websocket variables

mj23000 committed:
Add WebSocket connection test

mj23000 committed:
Merge branch 'bangolufsen' of https://github.com/mj23000/core into bangolufsen

mj23000 commented on homeassistant/components/bangolufsen/entity.py:
```diff
@@ -0,0 +1,74 @@
+        # get the input from the config entry.
+        self.entry: ConfigEntry = entry
+
+        self._device: DeviceEntry | None = None
```
I've removed self._device and now use device_entry.

mj23000 committed:
Remove unused constants

mj23000 commented on homeassistant/components/bangolufsen/const.py:
```diff
@@ -0,0 +1,224 @@
+from homeassistant.components.media_player import MediaPlayerState, MediaType
+
+
+class SOURCE_ENUM(StrEnum):
```
These values are the same as on the device, so they should not be translated (at least for now).

mj23000 committed:
Merge branch 'dev' into bangolufsen

mj23000 commented:
All of the requested changes should have been adressed. Thank you again for the feedback!

Kind regards

emontnemery commented on .strict-typing:
```diff
@@ -77,6 +77,7 @@ homeassistant.components.automation.*
 homeassistant.components.awair.*
 homeassistant.components.backup.*
 homeassistant.components.baf.*
+homeassistant.components.bang_olufsen.*
```
Nice!

emontnemery commented on homeassistant/components/bang_olufsen/config_flow.py:
```diff
@@ -0,0 +1,162 @@
+            try:
+                ipaddress.ip_address(self._host)
+            except ValueError:
+                return self.async_abort(reason="value_error")
```
If there's some mistake in the user input we should not abort the flow, instead we should ask the user to try again by showing the form again, with errors set, for example like so:

https://github.com/home-assistant/core/blob/98c98ab19901819b9d9f93f50c24675f2df57cfb/homeassistant/components/meater/config_flow.py#L79-L104

emontnemery commented on homeassistant/components/bang_olufsen/config_flow.py:
```diff
@@ -0,0 +1,162 @@
+
+            # Check if the IP address is a valid address.
+            try:
+                ipaddress.ip_address(self._host)
```
We normally allow either a hostname or an IP-address, is there some reason we only allow IP-addresses here?

emontnemery commented on homeassistant/components/bang_olufsen/config_flow.py:
```diff
@@ -0,0 +1,162 @@
+            await self.async_set_unique_id(self._serial_number)
+            self._abort_if_unique_id_configured()
+
+            return await self.async_step_confirm()
```
Why do we need a confirmation here?

mj23000 replied:
We do not. This was left over from when volume options were available. I've removed the confirm step entirely and Zeroconf discovered devices are now added to Home Assistant without user interaction.

mj23000 committed:
Merge branch 'dev' into bangolufsen

mj23000 committed:
Remove confirmation form
Make discovered devices get added to Home Assistant immediately
Fix device not being available on mdns discovery
Change config flow aborts to forms with error

mj23000 committed:
Merge branch 'bangolufsen' of https://github.com/mj23000/core into bangolufsen

mj23000 committed:
Update tests for new config_flow
Add missing api_exception test

mj23000 committed:
Restrict manual and discovered IP addresses to IPv4

emontnemery commented on homeassistant/components/bang_olufsen/strings.json:
```diff
@@ -1,19 +1,17 @@
-      "already_in_progress": "[%key:common::config_flow::abort::already_in_progress%]",
+    "error": {
       "api_exception": "[%key:common::config_flow::error::invalid_host%]",
       "client_connector_error": "[%key:common::config_flow::error::invalid_host%]",
```
Is the error message "Invalid hostname or IP address" really good for these issues?

mj23000 replied:
No it is not. I've changed it to "Failed to connect" from `string.json`.

emontnemery commented on homeassistant/components/bang_olufsen/strings.json:
```diff
@@ -1,19 +1,17 @@
       "api_exception": "[%key:common::config_flow::error::invalid_host%]",
       "client_connector_error": "[%key:common::config_flow::error::invalid_host%]",
       "timeout_error": "[%key:common::config_flow::error::timeout_connect%]",
       "value_error": "[%key:common::config_flow::error::invalid_host%]"
```
Since the API only allows actual IP addresses, maybe this integration should have it's own error message? Or is "Invalid hostname or IP address" good enough?

mj23000 replied:
Yes, I've changed the error message to be "Invalid IPv4 address" and the last commit added IPv4 address checking

emontnemery commented on homeassistant/components/bang_olufsen/config_flow.py:
```diff
@@ -118,45 +127,22 @@ async def async_step_zeroconf(
-        self.context["title_placeholders"] = {
-            "name": discovery_info.properties[ATTR_FRIENDLY_NAME]
-        }
+        return await self._create_entry()
```
We need a confirm step for devices discovered via zeroconf. If you search the codebase for `async_step_zeroconf_confirm` you will find many examples 👍

mj23000 replied:
Ah. so I did it correctly the first time. I've added a confirm step for zeroconf discovery.

emontnemery replied:
Yes, I'm sorry for the confusing review comments previously.

emontnemery commented on homeassistant/components/bang_olufsen/config_flow.py:
```diff
@@ -118,45 +127,22 @@ async def async_step_zeroconf(
+    async def _create_entry(self) -> FlowResult:
+        """Create the config entry for a discovered or manually configured Bang & Olufsen device."""
         await self.async_set_unique_id(self._serial_number)
         self._abort_if_unique_id_configured(updates={CONF_HOST: self._host})
```
Move this to `async_step_zeroconf`, this should happen without user confirmation

emontnemery commented on homeassistant/components/bang_olufsen/config_flow.py:
```diff
@@ -75,33 +86,31 @@ async def async_step_user(
+                        step_id="user",
+                        data_schema=data_schema,
+                        errors={
+                            "base": {
+                                ApiException: "api_exception",
+                                ClientConnectorError: "client_connector_error",
+                                TimeoutError: "timeout_error",
+                            }[type(error)]
```
Move this dictionary to module level

emontnemery commented on homeassistant/components/bang_olufsen/config_flow.py:
```diff
@@ -122,7 +122,13 @@ async def async_step_zeroconf(
+        # Ensure that an IPv4 address is received
         self._host = discovery_info.host
+        try:
+            IPv4Address(self._host)
```
Is this a limitation of the API?

mj23000 replied:
Yes.

mj23000 committed:
Re-add confirmation step for zeroconf discovery
Improve error messages
Move exception mapping dict to module level

mj23000 commented on homeassistant/components/bang_olufsen/config_flow.py:
```diff
@@ -64,9 +71,13 @@ async def async_step_user(
+                return self.async_show_form(
+                    step_id="user",
+                    data_schema=data_schema,
+                    errors={"base": "value_error"},
```
Yes, I've renamed it to invalid_ip.

mj23000 committed:
Merge branch 'dev' into bangolufsen

mj23000 committed:
Enable remote control WebSocket listener

mj23000 committed:
Update tests

emontnemery reviewed (commented):
Looks good, thanks @mj23000 👍
